File: scala/Quicksort.scala
import scala.io.Source

object Quicksort extends App {

  def quicksort(xs: List[Int]): List[Int] = xs match {
    case Nil => Nil
    case pivot :: tail =>
      val (lower, upper) = tail.partition(_ < pivot)
      quicksort(lower) ::: pivot :: quicksort(upper)
  }

  val xs = Source.stdin.mkString.split("\\s").drop(1).map(_.toInt).toList
  val sorted = quicksort(xs)
  println(sorted.mkString(" "))
}

File: kotlin/quicksort.kt
fun quicksort(xs: List<Int>): List<Int> =
    if (xs.isEmpty()) xs
    else {
        val pivot = xs.first()
        val (lower, upper) = xs.subList(1, xs.size).partition { it < pivot }
        quicksort(lower) + listOf(pivot) + quicksort(upper)
    }

fun main(args: Array<String>) {
    readLine()
    val xs = readLine()!!.split(" ").map { it.toInt() }.toList()
    val sorted = quicksort(xs)
    println(sorted.joinToString(" "))
}
